### PR TITLE
MOD-2418/2450 TI.Map iOS: Support rotation on annotations

### PR DESCRIPTION
**Pull request by anil-shukla-axway**
https://jira.appcelerator.org/browse/MOD-2418
https://jira.appcelerator.org/browse/MOD-2450

Commit 1dd8d9f is  just to revert changes of PR #297, which will reopen ticket [MOD-2103](https://jira.appcelerator.org/browse/MOD-2103) as the change caused a regression.

### CODE REVIEW

File: ios/titanium.xcconfig [hunk 1]
```diff
@@ -6,14 +6,9 @@
...
+//
 // THESE SHOULD BE OK GENERALLY AS-IS
 //
-TITANIUM_SDK = ~/Library/Application Support/Titanium/mobilesdk/osx/$(TITANIUM_SDK_VERSION)
-TITANIUM_BASE_SDK = "$(TITANIUM_SDK)/iphone/include"
-TITANIUM_BASE_SDK2 = "$(TITANIUM_SDK)/iphone/include/TiCore"
-TITANIUM_BASE_SDK3 = "$(TITANIUM_SDK)/iphone/include/ASI"
-TITANIUM_BASE_SDK4 = "$(TITANIUM_SDK)/iphone/include/APSHTTPClient"
-HEADER_SEARCH_PATHS = $(TITANIUM_BASE_SDK) $(TITANIUM_BASE_SDK2) $(TITANIUM_BASE_SDK3) $(TITANIUM_BASE_SDK4)
+TITANIUM_SDK = /Users/$(USER)/Library/Application Support/Titanium/mobilesdk/osx/$(TITANIUM_SDK_VERSION)
```
Comment (fasihul-axway) [inline ios/titanium.xcconfig:12]: @anil-shukla-axway, its better to separate config file changes into separate PR.

**Comment by fasihul-axway** [comment]
Test cases are missing, please add test case to verify the feature as mentioned on ticket.

File: ios/Classes/TiMapView.m [hunk 2]
```diff
@@ -1134,7 +1138,7 @@ - (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotationProxy:(TiMap
   if (customView == nil && !marker) {
     id imagePath = [ann valueForUndefinedKey:@"image"];
     image = [TiUtils image:imagePath proxy:ann];
-    identifier = (image != nil) ? @"timap-image" : @"timap-pin";
+    identifier = (image != nil) ? @"timap-image" : @"carIcon"; //carIcon
```
Comment (vikas-goyal-axway) [inline ios/Classes/TiMapView.m:1141]: @anil-shukla-axway , Why did you changed 'timap-pin' to `carIcon` ?

Comment (anil-shukla-axway) [reply ios/Classes/TiMapView.m]: Updated the name, Please verfiy.

File: ios/Classes/TiMapView.m [hunk 3]
```diff
@@ -1241,10 +1244,51 @@ - (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotationProxy:(TiMap
...
+
+  return 0.0f;
+}
+- (void)updateLoacation:(TiMapAnnotationProxy *)newAnnotation withLocation:(CLLocationCoordinate2D)newLocation
+{
+  MKCoordinateRegion regionInfo;
+  MKCoordinateSpan span;
+  regionInfo.span = span;
+  regionInfo.center = newLocation;
+  [self.map setRegion:regionInfo animated:YES];
```
Comment (vikas-goyal-axway) [inline ios/Classes/TiMapView.m:1278]: I don't think we should change region. It could cause  a problem if user set multiple location for multiple annotation.

Comment (anil-shukla-axway) [reply ios/Classes/TiMapView.m:1278]: This code is used to show user where the notification get lands, as once the annotation got move from screen user will never know where it goes. Do you think we should confirm this from Vijay?

File: ios/Classes/TiMapView.m [hunk 4]
```diff
@@ -1291,7 +1286,7 @@ - (void)updateLoacation:(TiMapAnnotationProxy *)newAnnotation withLocation:(CLLo
 }
 // mapView:viewForAnnotation: provides the view for each annotation.
 // This method may be called for all or some of the added annotations.
-// For MapKit provided annotations (eg. MKUserLocation) return nil to use the MapKit provided annotation view.
+// For MapKit provided annotations (eg. MKUserLocation) return nil to use the MapKit provided annotatiow2n view.
```
Comment (vikas-goyal-axway) [inline ios/Classes/TiMapView.m:1289]: @anil-shukla-axway , Please fix typing mistake

File: ios/Classes/TiMapView.m [hunk 5]
```diff
@@ -475,6 +478,7 @@ - (void)setUserLocation_:(id)value
     // else, just apply the userLocation
     [self map].showsUserLocation = userLocation;
   }
+  //[locationManager startUpdatingLocation];
```
Comment (vikas-goyal-axway) [inline ios/Classes/TiMapView.m:481]: @anil-shukla-axway , Remove unused code

File: ios/titanium.xcconfig (hunk 1 above)
Comment (vikas-goyal-axway) [reply ios/titanium.xcconfig:12]: @anil-shukla-axway , I agree with @fasihul-axway , Please address these changes in different PR

Comment (anil-shukla-axway) [reply ios/titanium.xcconfig]: Done, Please check PR - https://github.com/appcelerator-modules/ti.map/pull/296

File: ios/map.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -480,12 +480,6 @@
 				GCC_WARN_UNUSED_PARAMETER = NO;
 				GCC_WARN_UNUSED_VALUE = NO;
 				GCC_WARN_UNUSED_VARIABLE = NO;
-				HEADER_SEARCH_PATHS = (
```
Comment (vikas-goyal-axway) [inline ios/map.xcodeproj/project.pbxproj:483]: @anil-shukla-axway , Move these changes to https://github.com/appcelerator-modules/ti.map/pull/296

**Comment by vijaysingh-axway** [comment]
@anil-shukla-axway @vikas-goyal-axway @fasihul-axway I think we should introduce 2 new properties on annotation. I think the API design should be as following-

1. For rotation of annotation -  Ti.Map.annotation.rotate(angle) . We want something similar to android's rotation of marker https://developers.google.com/maps/documentation/android-sdk/marker#rotate_a_marker .
2. For animation of annotation- Ti.Map.annotation.animate(newLocation)

See following example -
Test Case - 
`var Map = require('ti.map');
var win = Ti.UI.createWindow({
    backgroundColor: 'white',
    title: 'Rotate'
});

var annotation = null;

var map = Map.createView({
    region: {
        latitude: 45,
        longitude: -76,
        latitudeDelta: 0.08,
        longitudeDelta: 0.08
    },
    bottom: 40
});
win.add(map);

var animateButton = Ti.UI.createButton({
    title: "Animate annotation to new location",
    bottom: 5
});
win.add(animateButton);
animateButton.addEventListener('click', function() {
    var newCoordinate = [45.05,-76.05];
    annotation.animate(newCoordinate);
});

var rotateButton = Ti.UI.createButton({
    title: "Rotate annotation to 90 degree",
    bottom: 5
});
win.add(rotateButton);

rotateButton.addEventListener('click', function() {
    annotation.rotate(90);
});

win.open();

createAnnotations();

function createAnnotations() {
    annotation = Map.createAnnotation({
            latitude: 45,
            longitude: -76,
            title: "annotation "
    });
    map.addAnnotation(annotation);
}`

Let me know if you have any question.

**Comment by anil-shukla-axway** [comment]
> @anil-shukla-axway @vikas-goyal-axway @fasihul-axway I think we should introduce 2 new properties on annotation -
> 
> 1. For rotation of annotation -  Ti.Map.annotation.rotate(angle) . We want something similar to android's rotation of marker https://developers.google.com/maps/documentation/android-sdk/marker#rotate_a_marker .
> 2. For animation of annotation- Ti.Map.annotation.animate(newLocation)
> 
> See following example -
> Test Case -
> `var Map = require('ti.map');
> var win = Ti.UI.createWindow({
> backgroundColor: 'white',
> title: 'Rotate'
> });
> 
> var annotation = null;
> 
> var map = Map.createView({
> region: {
> latitude: 45,
> longitude: -76,
> latitudeDelta: 0.08,
> longitudeDelta: 0.08
> },
> bottom: 100
> });
> win.add(map);
> 
> var animateButton = Ti.UI.createButton({
> title: "Animate annotation to new location",
> bottom: 5
> });
> win.add(animateButton);
> animateButton.addEventListener('click', function() {
> var newCoordinate = [45.05,-76.05];
> annotation.animate(newCoordinate);
> });
> 
> var rotateButton = Ti.UI.createButton({
> title: "Rotate annotation to 90 degree",
> bottom: 50
> });
> win.add(rotateButton);
> 
> rotateButton.addEventListener('click', function() {
> annotation.rotate(90);
> });
> 
> win.open();
> 
> createAnnotations();
> 
> function createAnnotations() {
> annotation = Map.createAnnotation({
> latitude: 45,
> longitude: -76,
> title: "annotation "
> });
> map.addAnnotation(annotation);
> }`
> 
> Let me know if you have any question.

I have followed the suggestions(Teams chat) from you.. I found that the methods of class TiMapAnnotationProxy are used to set properties of annotations below is example-  

var anno4 = Map.createAnnotation({
        latitude: -33.86365,
        longitude: 151.22689,
        title: 'Anno4',
        subtitle: 'This is anno4',
        draggable: true
    });

so If I create one method - (void)animate:(id)args; this has be like setting the animate([annotation, destLatLong]) at the time of annotation creation.

Also the below method --

- (void)setAnimate:(id)arg
{
  ENSURE_SINGLE_ARG(arg, NSArray);
  TiMapAnnotationProxy *newAnnotation = [arg objectAtIndex:0];
  NSArray *newLoc = [arg objectAtIndex:1];
  CLLocationCoordinate2D newLocation;
  newLocation.latitude = [[newLoc objectAtIndex:0] floatValue];
  newLocation.longitude = [[newLoc objectAtIndex:1] floatValue];

  [(TiMapView *)[self view] animateAnnotation:newAnnotation withLocation:newLocation];
}

here user has to provide annotation and destination lat long so on behalf of which an angle is being calculated and also the annotation is being animated.
So do we also have to provide a separate method to rotate the angle? as its been done by this below method already .. 


- (void)animateAnnotation:(TiMapAnnotationProxy *)newAnnotation withLocation:(CLLocationCoordinate2D)newLocation
{
  userOldLocation.latitude = newAnnotation.coordinate.latitude;
  userOldLocation.longitude = newAnnotation.coordinate.longitude;
  userNewLocation.latitude = newLocation.latitude;
  userNewLocation.longitude = newLocation.longitude;
  **float getAngle** = [self angleFromCoordinate:userOldLocation toCoordinate:userNewLocation];

  [UIView animateWithDuration:2
                   animations:^{
                     newAnnotation.coordinate = userNewLocation;
                     MKAnnotationView *annotationView = (MKAnnotationView *)[self.map viewForAnnotation:newAnnotation];
                     **annotationView.transform = CGAffineTransformMakeRotation(getAngle);**
                   }];
}

File: ios/example/tests/rotateAnnotation.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            longitude: -76,
+            title: "annotation "
+    });
+    map.addAnnotation(annotation);
+}
+
+function updateAnnotations() {
+    console.log("updateAnnotations");
+    var destLatLong = [45.05,-76.05];
+    map.animateAnnotation([annotation, destLatLong]);
```
Comment (vijaysingh-axway) [inline ios/example/tests/rotateAnnotation.js:40]: There is no API named 'animateAnnotation'. It is 'animate' on Annotation. Please verify test cases at your end.

File: ios/example/tests/annotations.js [hunk 8]
```diff
@@ -14,6 +14,22 @@ exports.run = function(UI, Map) {
...
+                var newLocation = [-33.09,151.03];
+                anno2.animate(newLocation);
+            }
+        },
+        {
+            hasChild: true,
+            title: 'rotate annotation',
+            run: function() {
+                var angle = 90;
+                anno4.rotationAngle(angle);
```
Comment (vijaysingh-axway) [inline ios/example/tests/annotations.js:30]:
```suggestion
                anno4.rotate(angle);
```

File: ios/example/tests/rotateAnnotation.js (hunk 7 above)
Comment (vijaysingh-axway) [reply ios/example/tests/rotateAnnotation.js:40]: Add a similar test case for rotation as well.

File: ios/Classes/TiMapAnnotationProxy.h [hunk 9]
```diff
@@ -32,15 +32,14 @@
...
 - (id)nativePinColor;
 - (BOOL)animatesDrop;
 - (void)setHidden:(id)value;
-
 - (UIView *)leftViewAccessory;
 - (UIView *)rightViewAccessory;
-
 - (int)tag;
+- (void)animate:(id)arg;
+- (void)rotationAngle:(id)arg;
```
[the remark this answers is not in the document]
Comment (anil-shukla-axway) [reply ios/Classes/TiMapAnnotationProxy.h:43]: Done

File: ios/Classes/TiMapAnnotationProxy.m [hunk 10]
```diff
@@ -548,5 +548,24 @@ - (int)tag
 {
   return tag;
 }
+- (void)rotationAngle:(id)arg
```
[the remark this answers is not in the document]
Comment (anil-shukla-axway) [reply ios/Classes/TiMapAnnotationProxy.m:551]: Done

File: ios/Classes/TiMapView.m [hunk 11]
```diff
@@ -1257,13 +1259,43 @@ - (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotationProxy:(TiMap
...
+  userOldLocation.longitude = newAnnotation.coordinate.longitude;
+  userNewLocation.latitude = newLocation.latitude;
+  userNewLocation.longitude = newLocation.longitude;
+  float getAngle = [self angleFromCoordinate:userOldLocation toCoordinate:userNewLocation];
+
+  [UIView animateWithDuration:2
+                   animations:^{
+                     newAnnotation.coordinate = userNewLocation;
+                     MKAnnotationView *annotationView = (MKAnnotationView *)[self.map viewForAnnotation:newAnnotation];
+                     annotationView.transform = CGAffineTransformMakeRotation(getAngle);
```
[the remark this answers is not in the document]
Comment (anil-shukla-axway) [reply ios/Classes/TiMapView.m:1293]: Done

File: ios/Classes/TiMapView.m [hunk 12]
```diff
@@ -1257,13 +1259,43 @@ - (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotationProxy:(TiMap
     ann.controllerPreviewing = [controller registerForPreviewingWithDelegate:previewingDelegate sourceView:annView];
 #endif
   }
-
   return annView;
 }
 
+- (float)angleFromCoordinate:(CLLocationCoordinate2D)first
```
[the remark this answers is not in the document]
Comment (anil-shukla-axway) [reply ios/Classes/TiMapView.m:1265]: method removed

File: ios/example/tests/annotations.js (hunk 8 above)
Comment (anil-shukla-axway) [reply ios/example/tests/annotations.js:30]: done

File: ios/example/tests/rotateAnnotation.js (hunk 7 above)
Comment (anil-shukla-axway) [reply ios/example/tests/rotateAnnotation.js:40]: I am added test cases in annotation.js file so I am removing this rotateAnnotation.js file as its no longer needed.

File: ios/Classes/TiMapView.m [hunk 13]
```diff
@@ -1070,7 +1069,7 @@ - (TiMapAnnotationProxy *)proxyForAnnotation:(MKAnnotationView *)pinview
   for (id annotation in [map annotations]) {
     if ([annotation isKindOfClass:[TiMapAnnotationProxy class]]) {
       if ([(TiMapAnnotationProxy *)annotation tag] == pinview.tag) {
-        [annotation setView:pinview];
+        //[annotation setView:pinview];
```
Comment (vijaysingh-axway) [inline ios/Classes/TiMapView.m:1072]: @anil-shukla-axway  Why this line commented out? It was introduced in PR #297 to fix https://jira.appcelerator.org/browse/MOD-2103.

Comment (anil-shukla-axway) [reply ios/Classes/TiMapView.m:1072]: work is in progress on this, as due to this #297 now the changes done in my PR #293 are getting crashed. So I need to find an alternative way for MOD-2103.

**Comment by anil-shukla-axway** [comment]
0c667367375eb6880be8d613e84aae05364daf8c will open ticket MOD-2103 as the previous changes were causing a regression.

**Comment by fasihul-axway** [comment]
@anil-shukla-axway, there are too many commit and some are useless and may cause confusion if someone dig and want to see history of changes. Please refactor and use specific commit with appropriate message. You have used `feat` in almost all commit which is for new feature, please use appropriate key.

**Comment by anil-shukla-axway** [comment]
> @anil-shukla-axway, there are too many commit and some are useless and may cause confusion if someone dig and want to see history of changes. Please refactor and use specific commit with appropriate message. You have used `feat` in almost all commit which is for new feature, please use appropriate key.

Made the required changes, please verify.

File: apidoc/Annotation.yml [hunk 14]
```diff
@@ -325,3 +325,21 @@ properties:
...
+
+methods:
+
+  - name: animate
+    summary: Animate annotation to new location.
+    parameters:
+        - name: newLocation
+          summary: latitude and longitude where annotation will animate.
+          type: Array<Number>
+    platforms: [iphone, ipad]
```
Comment (ewanharris) [inline apidoc/Annotation.yml:337]: A `since` property needs to be set on new APIs. This should be set to the SDK version we're expecting to ship the changes in.

**Review by ssjsamir — APPROVED** [review]
**FR Passed**, Able to rotate and animate annotations, tested using the ti.map example app. 

**Test environment**
```
MacOS Big Sur: 11.0 Beta 5
Xcode: 12.0 Beta 6
Java Version: 1.8.0_242
Android NDK: 21.3.6528147
Node.js: 12.18.1
""NPM":"5.0.0","CLI":"8.1.0-master.11""
iphone 8 Sim (14.0 Beta)
```

**Comment by ssjsamir** [comment]
@anil-shukla-axway Could you make updates to the package.json and manifest files so we can publish a new version of ti.map.

**Comment by satinder-singh-axway** [comment]
@anil-shukla-axway, Also please update the since property for YML change for newly added methods to 9.2.0 as this will be part of 9.2.0 release. Correct me @ssjsamir if wrong.
